File: backend/src/main/java/com/garah/api/catalogue/domaine/Attribut.java
package com.garah.api.catalogue.domaine;

import jakarta.persistence.*;

import java.util.ArrayList;
import java.util.List;

/**
 * Une dimension de déclinaison : Taille, Couleur, Capacité…
 *
 * <p>Le référentiel des attributs est <b>partagé par tout le catalogue</b> :
 * « Taille » est défini une fois et réutilisé par tous les vêtements. Sans ça,
 * chaque produit inventerait ses propres valeurs et aucun filtre transversal
 * ne serait possible.</p>
 */
@Entity
@Table(name = "attribut")
public class Attribut {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 40)
    private String code;

    @Column(nullable = false, length = 100)
    private String nom;

    /** LISTE (déroulante) ou PASTILLE (carré de couleur). */
    @Column(name = "type_affichage", nullable = false, length = 20)
    private String typeAffichage = "LISTE";

    @OneToMany(mappedBy = "attribut", fetch = FetchType.LAZY,
               cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ValeurAttribut> valeurs = new ArrayList<>();

    protected Attribut() {
    }

    public Attribut(String code, String nom, String typeAffichage) {
        this.code = code;
        this.nom = nom;
        this.typeAffichage = typeAffichage;
    }

    public ValeurAttribut ajouterValeur(String code, String libelle, String affichage) {
        ValeurAttribut valeur = new ValeurAttribut(this, code, libelle, affichage);
        valeurs.add(valeur);
        return valeur;
    }

    public Long getId() { return id; }
    public String getCode() { return code; }
    public String getNom() { return nom; }
    public String getTypeAffichage() { return typeAffichage; }
    public List<ValeurAttribut> getValeurs() { return valeurs; }
}

File: frontend/projects/garah-ui/src/lib/modeles/attribut.ts
/**
 * Une dimension de declinaison : Taille, Couleur, Conditionnement.
 *
 * C'est ce qu'Amazon appelle un « variation theme » : l'axe selon lequel un
 * produit se decline.
 *
 * Le referentiel est PARTAGE par tout le catalogue. « Taille » est defini une
 * fois et reutilise par tous les vetements — sinon chaque produit inventerait
 * ses propres valeurs (« M », « m », « Medium », « Moyen ») et aucun filtre
 * transversal ne serait possible.
 */
export interface Attribut {
  readonly id: number;
  /** Engendre depuis le nom, immuable. C'est lui qu'un import designe. */
  readonly code: string;
  readonly nom: string;
  readonly typeAffichage: TypeAffichage;
  readonly valeurs: readonly ValeurAttribut[];
}

/**
 * LISTE : une deroulante. PASTILLE : un carre de couleur.
 *
 * Une couleur ecrite en toutes lettres oblige le client a imaginer « Bleu
 * ciel ». La pastille la montre.
 */
export type TypeAffichage = 'LISTE' | 'PASTILLE';

export interface ValeurAttribut {
  readonly id: number;
  readonly code: string;
  readonly libelle: string;
  /** La couleur d'une pastille, en hexadecimal. Nulle sur une liste. */
  readonly valeurAffichage: string | null;
  /**
   * 38 avant 39 avant 40.
   *
   * Trie alphabetiquement, on obtient « 10, 38, 9 » : une taille se lit dans
   * l'ordre des tailles, pas dans celui du dictionnaire.
   */
  readonly ordre: number;
}

/** Les deux affichages proposes a la creation d'une dimension. */
export const TYPES_AFFICHAGE: readonly {
  readonly code: TypeAffichage;
  readonly libelle: string;
  readonly explication: string;
}[] = [
  {
    code: 'LISTE',
    libelle: 'Liste',
    explication: 'Pour les tailles, les capacites, les conditionnements.',
  },
  {
    code: 'PASTILLE',
    libelle: 'Pastille de couleur',
    explication: 'Pour les couleurs : le client voit la teinte, il ne l’imagine pas.',
  },
];

/**
 * Une valeur retenue par une declinaison, telle que l'API la renvoie.
 *
 * Porte de quoi l'afficher sans relire le referentiel : le nom de sa
 * dimension et, le cas echeant, sa couleur.
 */
export interface ValeurChoisie {
  readonly valeurId: number;
  readonly attributId: number;
  readonly attribut: string;
  readonly libelle: string;
  readonly valeurAffichage: string | null;
}
